File: server/src/main/java/uk/co/unclealex/hammers/calendar/server/html/GameUpdateCommand.java
/**
 * Copyright 2011 Alex Jones
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.    
 *
 * @author unclealex72
 *
 */

package uk.co.unclealex.hammers.calendar.server.html;

import org.joda.time.DateTime;

import uk.co.unclealex.hammers.calendar.server.model.Game;

/**
 * 
 * A class used to encapsulate a possible update to a game as directed by an
 * {@link HtmlGamesScanner}.
 * 
 * @author alex
 * 
 */
public abstract class GameUpdateCommand implements Comparable<GameUpdateCommand> {

	/**
	 * Create a {@link GameUpdateCommand} that updates a game's date played value.
	 * 
	 * @param gameLocator
	 *          The locator to use to locate the game.
	 * @param newDatePlayed
	 *          The new date played value.
	 * @return A {@link GameUpdateCommand} that updates a game's date played
	 *         value.
	 */
	public static GameUpdateCommand datePlayed(GameLocator gameLocator, DateTime newDatePlayed) {
		return new InternalGameUpdateCommand<DateTime>(Type.DATE_PLAYED, gameLocator, newDatePlayed) {
			@Override
			protected DateTime getCurrentValue(Game game) {
				return game.getDatePlayed();
			}

			@Override
			protected void setNewValue(Game game, DateTime newDatePlayed) {
				game.setDatePlayed(newDatePlayed);
			}
		};
	}

	/**
	 * Create a {@link GameUpdateCommand} that updates a game's result value.
	 * 
	 * @param gameLocator
	 *          The locator to use to locate the game.
	 * @param newResult
	 *          The new result.
	 * @return A {@link GameUpdateCommand} that updates a game's result value.
	 */
	public static GameUpdateCommand result(GameLocator gameLocator, String newResult) {
		return new InternalGameUpdateCommand<String>(Type.RESULT, gameLocator, newResult) {
			@Override
			protected String getCurrentValue(Game game) {
				return game.getResult();
			}

			@Override
			protected void setNewValue(Game game, String newResult) {
				game.setResult(newResult);
			}
		};
	}

	/**
	 * Create a {@link GameUpdateCommand} that updates a game's attendence value.
	 * 
	 * @param gameLocator
	 *          The locator to use to locate the game.
	 * @param newAttendence
	 *          The new attendence.
	 * @return A {@link GameUpdateCommand} that updates a game's attendence value.
	 */
	public static GameUpdateCommand attendence(GameLocator gameLocator, Integer newAttendence) {
		return new InternalGameUpdateCommand<Integer>(Type.ATTENDENCE, gameLocator, newAttendence) {
			@Override
			protected Integer getCurrentValue(Game game) {
				return game.getAttendence();
			}

			@Override
			protected void setNewValue(Game game, Integer newAttendence) {
				game.setAttendence(newAttendence);
			}
		};
	}

	/**
	 * Create a {@link GameUpdateCommand} that updates a game's match report
	 * value.
	 * 
	 * @param gameLocator
	 *          The locator to use to locate the game.
	 * @param newMatchReport
	 *          The new match report.
	 * @return A {@link GameUpdateCommand} that updates a game's match report
	 *         value.
	 */
	public static GameUpdateCommand matchReport(GameLocator gameLocator, String newMatchReport) {
		return new InternalGameUpdateCommand<String>(Type.MATCH_REPORT, gameLocator, newMatchReport) {
			@Override
			protected String getCurrentValue(Game game) {
				return game.getMatchReport();
			}

			@Override
			protected void setNewValue(Game game, String newMatchReport) {
				game.setMatchReport(newMatchReport);
			}
		};
	}

	/**
	 * Create a {@link GameUpdateCommand} that updates a game's television channel
	 * value.
	 * 
	 * @param gameLocator
	 *          The locator to use to locate the game.
	 * @param newTelevisionChannel
	 *          The new television channel.
	 * @return A {@link GameUpdateCommand} that updates a game's television channel
	 *         value.
	 */
	public static GameUpdateCommand televisionChannel(GameLocator gameLocator, String newTelevisionChannel) {
		return new InternalGameUpdateCommand<String>(Type.TELEVISION_CHANNEL, gameLocator, newTelevisionChannel) {
			@Override
			protected String getCurrentValue(Game game) {
				return game.getTelevisionChannel();
			}

			@Override
			protected void setNewValue(Game game, String newTelevisionChannel) {
				game.setTelevisionChannel(newTelevisionChannel);
			}
		};
	}

	/**
	 * Create a {@link GameUpdateCommand} that updates a game's attended value.
	 * 
	 * @param gameLocator
	 *          The locator to use to locate the game.
	 * @param newAttended
	 *          The new attended value.
	 * @return A {@link GameUpdateCommand} that updates a game's attended value.
	 */
	public static GameUpdateCommand attended(GameLocator gameLocator, Boolean newAttended) {
		return new InternalGameUpdateCommand<Boolean>(Type.ATTENDED, gameLocator, newAttended) {
			@Override
			protected Boolean getCurrentValue(Game game) {
				return game.isAttended();
			}

			@Override
			protected void setNewValue(Game game, Boolean newAttended) {
				game.setAttended(newAttended);
			}
		};
	}

	/**
	 * Create a {@link GameUpdateCommand} that updates a game's bondholder tickets availability date.
	 * 
	 * @param gameLocator
	 *          The locator to use to locate the game.
	 * @param newBondHolderTicketsAvailable
	 *          The new date for bond holder ticket availabilty.
	 * @return A {@link GameUpdateCommand} that updates a game's bondholder availability tickets date.
	 */
	public static GameUpdateCommand bondHolderTickets(GameLocator gameLocator, DateTime newBondHolderTicketsAvailable) {
		return new InternalGameUpdateCommand<DateTime>(Type.BONDHOLDER_TICKETS, gameLocator, newBondHolderTicketsAvailable) {
			@Override
			protected DateTime getCurrentValue(Game game) {
				return game.getBondholdersAvailable();
			}

			@Override
			protected void setNewValue(Game game, DateTime newBondHolderTicketsAvailable) {
				game.setBondholdersAvailable(newBondHolderTicketsAvailable);
			}
		};
	}

	/**
	 * Create a {@link GameUpdateCommand} that updates a game's priority points tickets availability date.
	 * 
	 * @param gameLocator
	 *          The locator to use to locate the game.
	 * @param newPriorityPointTicketsAvailable
	 *          The new date for priority point ticket availabilty.
	 * @return A {@link GameUpdateCommand} that updates a game's priority points tickets availability date.
	 */
	public static GameUpdateCommand priorityPointTickets(GameLocator gameLocator, DateTime newPriorityPointTicketsAvailable) {
		return new InternalGameUpdateCommand<DateTime>(Type.PRIORITY_POINT_POST_TICKETS, gameLocator, newPriorityPointTicketsAvailable) {
			@Override
			protected DateTime getCurrentValue(Game game) {
				return game.getPriorityPointPostAvailable();
			}

			@Override
			protected void setNewValue(Game game, DateTime newPriorityPointTicketsAvailable) {
				game.setPriorityPointPostAvailable(newPriorityPointTicketsAvailable);
			}
		};
	}

	/**
	 * Create a {@link GameUpdateCommand} that updates a game's season tickets availability date.
	 * 
	 * @param gameLocator
	 *          The locator to use to locate the game.
	 * @param newSeasonTicketsAvailable
	 *          The new date for season ticket availabilty.
	 * @return A {@link GameUpdateCommand} that updates a game's season ticket availability date.
	 */
	public static GameUpdateCommand seasonTickets(GameLocator gameLocator, DateTime newSeasonTicketsAvailable) {
		return new InternalGameUpdateCommand<DateTime>(Type.SEASON_TICKETS, gameLocator, newSeasonTicketsAvailable) {
			@Override
			protected DateTime getCurrentValue(Game game) {
				return game.getSeasonTicketsAvailable();
			}

			@Override
			protected void setNewValue(Game game, DateTime newSeasonTicketsAvailable) {
				game.setSeasonTicketsAvailable(newSeasonTicketsAvailable);
			}
		};
	}

	/**
	 * Create a {@link GameUpdateCommand} that updates a game's academy tickets availability date.
	 * 
	 * @param gameLocator
	 *          The locator to use to locate the game.
	 * @param newAcademyTicketsAvailable
	 *          The new date for academy ticket availabilty.
	 * @return A {@link GameUpdateCommand} that updates a game's academy ticket availability date.
	 */
	public static GameUpdateCommand academyTickets(GameLocator gameLocator, DateTime newAcademyTicketsAvailable) {
		return new InternalGameUpdateCommand<DateTime>(Type.ACADEMY_TICKETS, gameLocator, newAcademyTicketsAvailable) {
			@Override
			protected DateTime getCurrentValue(Game game) {
				return game.getAcademyMembersAvailable();
			}

			@Override
			protected void setNewValue(Game game, DateTime newAcademyAvailable) {
				game.setAcademyMembersAvailable(newAcademyAvailable);
			}
		};
	}

	/**
	 * Create a {@link GameUpdateCommand} that updates a game's general sale tickets availability date.
	 * 
	 * @param gameLocator
	 *          The locator to use to locate the game.
	 * @param newGeneralSaleTicketsAvailable
	 *          The new date for general sale ticket availabilty.
	 * @return A {@link GameUpdateCommand} that updates a game's general sale ticket availability date.
	 */
	public static GameUpdateCommand generalSaleTickets(GameLocator gameLocator, DateTime newGeneralSaleTicketsAvailable) {
		return new InternalGameUpdateCommand<DateTime>(Type.GENERAL_SALE_TICKETS, gameLocator, newGeneralSaleTicketsAvailable) {
			@Override
			protected DateTime getCurrentValue(Game game) {
				return game.getGeneralSaleAvailable();
			}

			@Override
			protected void setNewValue(Game game, DateTime newAcademyAvailable) {
				game.setGeneralSaleAvailable(newAcademyAvailable);
			}
		};
	}

	/**
	 * An enumeration to allow the difference {@link GameUpdateCommand}s to be ordered in a sorted set.
	 * @author alex
	 *
	 */
	static enum Type {
		DATE_PLAYED, RESULT, ATTENDENCE, MATCH_REPORT, TELEVISION_CHANNEL, ATTENDED, 
		BONDHOLDER_TICKETS, PRIORITY_POINT_POST_TICKETS, SEASON_TICKETS, ACADEMY_TICKETS, GENERAL_SALE_TICKETS
	}

	private final GameLocator i_gameLocator;
	private final Type i_type;

	protected GameUpdateCommand(Type type, GameLocator gameLocator) {
		super();
		i_gameLocator = gameLocator;
		i_type = type;
	}

	/**
	 * Update a game. No check is made to see if the correct game is being updated.
	 * @param game The game to update.
	 * @return True if the game was updated, false otherwise.
	 */
	public abstract boolean update(Game game);
	
	@Override
	public int compareTo(GameUpdateCommand o) {
		int cmp = getGameLocator().compareTo(o.getGameLocator());
		if (cmp == 0) {
			cmp = getType().compareTo(o.getType());
		}
		return cmp;
	}

	@Override
	public boolean equals(Object obj) {
		return obj instanceof GameUpdateCommand && compareTo((GameUpdateCommand) obj) == 0;
	}
	
	/**
	 * A subclass of {@link GameUpdateCommand} that actually attempts to update a game.
	 * @author alex
	 *
	 * @param <V> The type of the value that may be updated.
	 */
	protected static abstract class InternalGameUpdateCommand<V> extends GameUpdateCommand {

		private final V i_newValue;

		protected InternalGameUpdateCommand(Type type, GameLocator gameLocator, V newValue) {
			super(type, gameLocator);
			i_newValue = newValue;
		}

		@Override
		public String toString() {
			return String.format("{%s: %s <- %s}", getGameLocator(), getType(), getNewValue());
		}
		
		public boolean update(Game game) {
			if (getNewValue() == null || getNewValue().equals(getCurrentValue(game))) {
				return false;
			}
			setNewValue(game, getNewValue());
			return true;
		}

		/**
		 * @param game The game to check.
		 * @return The current value of the current game.
		 */
		protected abstract V getCurrentValue(Game game);

		/**
		 * Alter a game.
		 * @param game The game to alter.
		 * @param newValue The new value to assign to the game.
		 */
		protected abstract void setNewValue(Game game, V newValue);

		/**
		 * @return the newValue
		 */
		public V getNewValue() {
			return i_newValue;
		}
	}

	/**
	 * @return The required {@link GameLocator} of any game to change.
	 */
	public GameLocator getGameLocator() {
		return i_gameLocator;
	}

	/**
	 * @return The ordering of this {@link GameUpdateCommand}.
	 */
	protected Type getType() {
		return i_type;
	}
}

File: app/html/GameUpdateCommand.scala
/**
 * Copyright 2010-2012 Alex Jones
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with work for additional information
 * regarding copyright ownership.  The ASF licenses file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */
package html

import com.typesafe.scalalogging.slf4j.StrictLogging
import enumeratum.{EnumEntry, _}
import model.Game
import models.GameResult
import org.joda.time.DateTime

/**
 * A class used to encapsulate a possible update to a game as directed by an.
 *
 * {@link HtmlGamesScanner}.
 *
 * @author alex
 */
sealed abstract class GameUpdateCommand(
  /**
   * The locator to use to locate the game.
   */
  val gameLocator: GameLocator) {

  /**
   * Update a game. No check is made to see if the correct game is being
   * updated.
   *
   * @param game
   *          The game to update.
   * @return The updated game if it was updated or none otherwise.
   */
  def update(game: Game): Option[Game]
}

sealed abstract class UpdateType(val description: String) extends EnumEntry
object UpdateType extends Enum[UpdateType] {

  val values = findValues

  // The type for date played updates.
  case object DATE_PLAYED extends UpdateType("date played")
  // The type for result updates.
  case object RESULT extends UpdateType("result")
  // The type for attendence updates.
  case object ATTENDENCE extends UpdateType("attendence")
  // The type for match report updates.
  case object MATCH_REPORT extends UpdateType("match report")
  // The type for television channel updates.
  case object TELEVISION_CHANNEL extends UpdateType("television channel")
  // The type for attendence updates.
  case object ATTENDED extends UpdateType("attendance flag")
  // The type for home team image updates.
  case object HOME_TEAM_IMAGE_LINK extends UpdateType("home team image")
  // The type for away team image updates.
  case object AWAY_TEAM_IMAGE_LINK extends UpdateType("away team image")
  // The type for competition image updates.
  case object COMPETITION_IMAGE_LINK extends UpdateType("competition image")
  // The type for changing the Bondholder ticket selling dates.
  case object BONDHOLDER_TICKETS extends UpdateType("bondholder ticket selling date")
  // The type for changing the priority point ticket selling dates.
  case object PRIORITY_POINT_POST_TICKETS extends UpdateType("priority ticket selling date")
  // The type for changing the season ticket holder ticket selling dates.
  case object SEASON_TICKETS extends UpdateType("season ticket selling date")
  // The type for changing the Academy members' ticket selling dates.
  case object ACADEMY_TICKETS extends UpdateType("academy ticket selling date")
  // The type for changing the general sale ticket selling dates.
  case object GENERAL_SALE_TICKETS extends UpdateType("general sale ticket selling date")

  implicit val ordering: Ordering[UpdateType] = Ordering.by(values.indexOf)
}

import html.UpdateType._

sealed abstract class BaseGameUpdateCommand[V](
  /**
   * The type of update (used for ordering).
   */
  val updateType: UpdateType,
  /**
   * The locator to use to locate the game.
   */
  override val gameLocator: GameLocator,
  /**
   * The new date played value.
   */
  val newValue: V) extends GameUpdateCommand(gameLocator) with StrictLogging {

  override def update(game: Game): Option[Game] = {
    if (!currentValue(game).contains(newValue)) {
      logger info s"Updating the ${updateType.description} to $newValue for game ${game.gameKey}"
      Some(setNewValue(game))
    } else {
      None
    }
  }

  /**
   * Gets the current value.
   *
   * @param game
   *          The game to check.
   * @return The current value of the current game.
   */
  protected def currentValue(game: Game): Option[V]

  /**
   * Alter a game.
   *
   * @param game
   *          The game to alter.
   */
  protected def setNewValue(game: Game): Game
}

/**
 * A {@link GameUpdateCommand} that updates a game's date played value.
 *
 * @param gameLocator
 *          The locator to use to locate the game.
 * @param newValue
 *          The new date played value.
 * @return A {@link GameUpdateCommand} that updates a game's date played
 *         value.
 */
case class DatePlayedUpdateCommand(
  override val gameLocator: GameLocator, override val newValue: DateTime) extends BaseGameUpdateCommand[DateTime](DATE_PLAYED, gameLocator, newValue) {

  override def currentValue(game: Game) = game.at

  override def setNewValue(game: Game) = game.copy(at = Some(newValue))
}

/**
 * Create a {@link GameUpdateCommand} that updates a game's result value.
 *
 * @param gameLocator
 *          The locator to use to locate the game.
 * @param newValue
 *          The new result.
 * @return A {@link GameUpdateCommand} that updates a game's result value.
 */
case class ResultUpdateCommand(
  override val gameLocator: GameLocator, override val newValue: GameResult) extends BaseGameUpdateCommand[GameResult](RESULT, gameLocator, newValue) {

  override def currentValue(game: Game) = game.result

  override def setNewValue(game: Game) = game.copy(result = Some(newValue))
}

/**
 * Create a {@link GameUpdateCommand} that updates a game's attendence value.
 *
 * @param gameLocator
 *          The locator to use to locate the game.
 * @param newValue
 *          The new attendence.
 * @return A {@link GameUpdateCommand} that updates a game's attendence value.
 */
case class AttendenceUpdateCommand(
  override val gameLocator: GameLocator, override val newValue: Int) extends BaseGameUpdateCommand[Int](ATTENDENCE, gameLocator, newValue) {

  override def currentValue(game: Game) = game.attendance

  override def setNewValue(game: Game) = game.copy(attendance = Some(newValue))

}

/**
  * Create a {@link GameUpdateCommand} that updates a game's home team image link value.
  *
  * @param gameLocator
  *          The locator to use to locate the game.
  * @param newValue
  *          The new home team image url.
  */
case class HomeTeamImageLinkCommand(
                                    override val gameLocator: GameLocator,
                                    override val newValue: String) extends BaseGameUpdateCommand[String](HOME_TEAM_IMAGE_LINK, gameLocator, newValue) {

  override def currentValue(game: Game) = game.homeTeamImageLink

  override def setNewValue(game: Game) = game.copy(homeTeamImageLink = Some(newValue))

}

/**
  * Create a {@link GameUpdateCommand} that updates a game's away team image link value.
  *
  * @param gameLocator
  *          The locator to use to locate the game.
  * @param newValue
  *          The new away team image url.
  */
case class AwayTeamImageLinkCommand(
                                     override val gameLocator: GameLocator,
                                     override val newValue: String) extends BaseGameUpdateCommand[String](AWAY_TEAM_IMAGE_LINK, gameLocator, newValue) {

  override def currentValue(game: Game) = game.awayTeamImageLink

  override def setNewValue(game: Game) = game.copy(awayTeamImageLink = Some(newValue))

}

/**
  * Create a {@link GameUpdateCommand} that updates a game's away team image link value.
  *
  * @param gameLocator
  *          The locator to use to locate the game.
  * @param newValue
  *          The new away team image url.
  */
case class CompetitionImageLinkCommand(
                                     override val gameLocator: GameLocator,
                                     override val newValue: String) extends BaseGameUpdateCommand[String](COMPETITION_IMAGE_LINK, gameLocator, newValue) {

  override def currentValue(game: Game) = game.competitionImageLink

  override def setNewValue(game: Game) = game.copy(competitionImageLink = Some(newValue))

}

/**
 * Create a {@link GameUpdateCommand} that updates a game's match report
 * value.
 *
 * @param gameLocator
 *          The locator to use to locate the game.
 * @param newValue
 *          The new match report.
 * @return A {@link GameUpdateCommand} that updates a game's match report
 *         value.
 */
case class MatchReportUpdateCommand(
  override val gameLocator: GameLocator, override val newValue: String) extends BaseGameUpdateCommand[String](MATCH_REPORT, gameLocator, newValue) {

  override def currentValue(game: Game) = game.matchReport

  override def setNewValue(game: Game) = game.copy(matchReport = Some(newValue))
}

/**
 * Create a {@link GameUpdateCommand} that updates a game's television channel
 * value.
 *
 * @param gameLocator
 *          The locator to use to locate the game.
 * @param newValue
 *          The new television channel.
 * @return A {@link GameUpdateCommand} that updates a game's television
 *         channel value.
 */
case class TelevisionChannelUpdateCommand(
  override val gameLocator: GameLocator, override val newValue: String) extends BaseGameUpdateCommand[String](TELEVISION_CHANNEL, gameLocator, newValue) {

  override def currentValue(game: Game) = game.televisionChannel

  override def setNewValue(game: Game) = game.copy(televisionChannel = Some(newValue))
}

/**
 * Create a {@link GameUpdateCommand} that updates a game's attended value.
 *
 * @param gameLocator
 *          The locator to use to locate the game.
 * @param newValue
 *          The new attended value.
 * @return A {@link GameUpdateCommand} that updates a game's attended value.
 */
case class AttendedUpdateCommand(
  override val gameLocator: GameLocator, override val newValue: Boolean) extends BaseGameUpdateCommand[Boolean](ATTENDED, gameLocator, newValue) {

  override def currentValue(game: Game) = Some(game.attended)

  override def setNewValue(game: Game) = game.copy(attended = newValue)
}

/**
 * The parent class of all ticketing update commands. This basically restricts the value type to be updated to DateTime.
 */
sealed abstract class TicketsUpdateCommand(
  override val updateType: UpdateType, override val gameLocator: GameLocator, override val newValue: DateTime)
  extends BaseGameUpdateCommand[DateTime](updateType, gameLocator, newValue)

/**
 * Create a {@link GameUpdateCommand} that updates a game's bondholder tickets
 * availability date.
 *
 * @param gameLocator
 *          The locator to use to locate the game.
 * @param newValue
 *          The new date for bond holder ticket availabilty.
 * @return A {@link GameUpdateCommand} that updates a game's bondholder
 *         availability tickets date.
 */
case class BondHolderTicketsUpdateCommand(
  override val gameLocator: GameLocator, override val newValue: DateTime) extends TicketsUpdateCommand(BONDHOLDER_TICKETS, gameLocator, newValue) {

  override def currentValue(game: Game) = game.bondholdersAvailable
  override def setNewValue(game: Game) = game.copy(bondholdersAvailable = Some(newValue))
}

/**
 * Create a {@link GameUpdateCommand} that updates a game's priority points
 * tickets availability date.
 *
 * @param gameLocator
 *          The locator to use to locate the game.
 * @param newValue
 *          The new date for priority point ticket availabilty.
 * @return A {@link GameUpdateCommand} that updates a game's priority points
 *         tickets availability date.
 */
case class PriorityPointTicketsUpdateCommand(
  override val gameLocator: GameLocator, override val newValue: DateTime) extends TicketsUpdateCommand(PRIORITY_POINT_POST_TICKETS, gameLocator, newValue) {

  override def currentValue(game: Game) = game.priorityPointAvailable
  override def setNewValue(game: Game) = game.copy(priorityPointAvailable = Some(newValue))
}

/**
 * Create a {@link GameUpdateCommand} that updates a game's season tickets
 * availability date.
 *
 * @param gameLocator
 *          The locator to use to locate the game.
 * @param newValue
 *          The new date for season ticket availabilty.
 * @return A {@link GameUpdateCommand} that updates a game's season ticket
 *         availability date.
 */
case class SeasonTicketsUpdateCommand(
  override val gameLocator: GameLocator, override val newValue: DateTime) extends TicketsUpdateCommand(SEASON_TICKETS, gameLocator, newValue) {

  override def currentValue(game: Game) = game.seasonTicketsAvailable
  override def setNewValue(game: Game) = game.copy(seasonTicketsAvailable = Some(newValue))
}

/**
 * Create a {@link GameUpdateCommand} that updates a game's academy tickets
 * availability date.
 *
 * @param gameLocator
 *          The locator to use to locate the game.
 * @param newValue
 *          The new date for academy ticket availabilty.
 * @return A {@link GameUpdateCommand} that updates a game's academy ticket
 *         availability date.
 */
case class AcademyTicketsUpdateCommand(
  override val gameLocator: GameLocator, override val newValue: DateTime) extends TicketsUpdateCommand(ACADEMY_TICKETS, gameLocator, newValue) {

  override def currentValue(game: Game) = game.academyMembersAvailable
  override def setNewValue(game: Game) = game.copy(academyMembersAvailable = Some(newValue))

}

/**
 * Create a {@link GameUpdateCommand} that updates a game's general sale
 * tickets availability date.
 *
 * @param gameLocator
 *          The locator to use to locate the game.
 * @param newValue
 *          The new date for general sale ticket availabilty.
 * @return A {@link GameUpdateCommand} that updates a game's general sale
 *         ticket availability date.
 */
case class GeneralSaleTicketsUpdateCommand(
  override val gameLocator: GameLocator, override val newValue: DateTime) extends TicketsUpdateCommand(GENERAL_SALE_TICKETS, gameLocator, newValue) {

  override def currentValue(game: Game) = game.generalSaleAvailable
  override def setNewValue(game: Game) = game.copy(generalSaleAvailable = Some(newValue))
}
